File: shared/permission_policy.py
from __future__ import annotations

from typing import Any

from platforms.feishu.config import FEISHU_APP_ID
from shared.permission_policy_loader import ALL, PermissionPolicyError, clean_text, load_permission_policy


_POLICY = load_permission_policy()

USER_LYX = _POLICY.user_name_to_union_id.get("LYX", "")
USER_ZQY = _POLICY.user_name_to_union_id.get("ZQY", "")
USER_ZGL = _POLICY.user_name_to_union_id.get("ZGL", "")
USER_AMAZON_REPLENISH_GROUP_1_MEMBER = _POLICY.user_name_to_union_id.get(
    "AMAZON_REPLENISH_GROUP_1_MEMBER",
    "",
)
USER_AMAZON_REPLENISH_GROUP_2_MEMBER = _POLICY.user_name_to_union_id.get(
    "AMAZON_REPLENISH_GROUP_2_MEMBER",
    "",
)
USER_AMAZON_REPLENISH_GROUP_3_MEMBER = _POLICY.user_name_to_union_id.get(
    "AMAZON_REPLENISH_GROUP_3_MEMBER",
    "",
)
USER_DEV_GROUP_MEMBER = _POLICY.user_name_to_union_id.get("DEV_GROUP_MEMBER", "")

BOT_LXE_CLAW = _POLICY.bot_alias_to_key.get("LXE_CLAW", "")
BOT_LXE_FBA_AGENT = _POLICY.bot_alias_to_key.get("AMAZON_FBA", "")
BOT_AMAZON_REPLENISH = _POLICY.bot_alias_to_key.get("AMAZON_REPLENISH", "")
BOT_AMAZON_REPLENISH_GROUP_2 = _POLICY.bot_alias_to_key.get("AMAZON_REPLENISH_GROUP_2", "")
BOT_AMAZON_REPLENISH_GROUP_3 = _POLICY.bot_alias_to_key.get("AMAZON_REPLENISH_GROUP_3", "")

BOT_ID_LXE_CLAW = _POLICY.bot_alias_to_app_id.get("LXE_CLAW", "")
BOT_ID_LXE_FBA_AGENT = _POLICY.bot_alias_to_app_id.get("AMAZON_FBA", "")
BOT_ID_AMAZON_REPLENISH = _POLICY.bot_alias_to_app_id.get("AMAZON_REPLENISH", "")
BOT_ID_AMAZON_REPLENISH_GROUP_2 = _POLICY.bot_alias_to_app_id.get("AMAZON_REPLENISH_GROUP_2", "")
BOT_ID_AMAZON_REPLENISH_GROUP_3 = _POLICY.bot_alias_to_app_id.get("AMAZON_REPLENISH_GROUP_3", "")

SKILL_TYPE_AMAZON_FBA = "amazon_fba"
SKILL_TYPE_AMAZON_REPLENISH = "amazon_replenish"
SKILL_TYPE_DEFAULT = "default"

BOT_ID_TO_KEY = dict(_POLICY.bot_id_to_key)
USER_AGENT_POLICY = {user_id: set(allowed) for user_id, allowed in _POLICY.user_agent_policy.items()}
BOT_SKILL_POLICY = {bot_key: set(skill_types) for bot_key, skill_types in _POLICY.bot_skill_policy.items()}


def _raw_data(source: Any) -> dict[str, Any]:
    raw = getattr(source, "raw_data", None)
    if isinstance(raw, dict):
        return dict(raw)
    raw = getattr(source, "source", None)
    if isinstance(raw, dict):
        return dict(raw)
    return {}


def _extra_data(source: Any) -> dict[str, Any]:
    raw = _raw_data(source)
    extra = raw.get("extra")
    return dict(extra) if isinstance(extra, dict) else {}


def _source_text(source: Any, name: str) -> str:
    try:
        return clean_text(getattr(source, name))
    except Exception:
        return ""


def bot_key_for_bot_id(bot_id: str) -> str:
    return BOT_ID_TO_KEY.get(clean_text(bot_id), "")


def is_known_bot_id(bot_id: str) -> bool:
    return bool(bot_key_for_bot_id(bot_id))


def resolve_permission_user_id(source: Any) -> str:
    raw = _raw_data(source)
    return (
        _source_text(source, "union_id")
        or clean_text(raw.get("union_id"))
        or clean_text(raw.get("sender_union_id"))
    )


def can_user_access_bot(user_id: str, bot_id: str) -> bool:
    bot_key = bot_key_for_bot_id(bot_id)
    if not bot_key:
        return False
    allowed = set(USER_AGENT_POLICY.get(clean_text(user_id), set()))
    return ALL in allowed or bot_key in allowed


def allowed_skill_types_for_bot(bot_id: str) -> set[str]:
    bot_key = bot_key_for_bot_id(bot_id)
    if not bot_key:
        return set()
    return set(BOT_SKILL_POLICY.get(bot_key, set()))


def resolve_bot_id(source: Any) -> str:
    raw = _raw_data(source)
    extra = _extra_data(source)
    direct_bot_id = (
        clean_text(raw.get("bot_id"))
        or clean_text(raw.get("app_id"))
        or clean_text(raw.get("bot_app_id"))
        or clean_text(extra.get("bot_app_id"))
        or clean_text(extra.get("bot_id"))
    )
    platform = (_source_text(source, "platform") or clean_text(raw.get("platform"))).lower()
    if platform == "feishu":
        return direct_bot_id or clean_text(FEISHU_APP_ID)

    return direct_bot_id


__all__ = [
    "ALL",
    "BOT_AMAZON_REPLENISH",
    "BOT_AMAZON_REPLENISH_GROUP_2",
    "BOT_AMAZON_REPLENISH_GROUP_3",
    "BOT_ID_AMAZON_REPLENISH",
    "BOT_ID_AMAZON_REPLENISH_GROUP_2",
    "BOT_ID_AMAZON_REPLENISH_GROUP_3",
    "BOT_ID_LXE_CLAW",
    "BOT_ID_LXE_FBA_AGENT",
    "BOT_LXE_CLAW",
    "BOT_LXE_FBA_AGENT",
    "BOT_SKILL_POLICY",
    "PermissionPolicyError",
    "SKILL_TYPE_AMAZON_FBA",
    "SKILL_TYPE_AMAZON_REPLENISH",
    "SKILL_TYPE_DEFAULT",
    "USER_AGENT_POLICY",
    "USER_AMAZON_REPLENISH_GROUP_1_MEMBER",
    "USER_AMAZON_REPLENISH_GROUP_2_MEMBER",
    "USER_AMAZON_REPLENISH_GROUP_3_MEMBER",
    "USER_DEV_GROUP_MEMBER",
    "USER_LYX",
    "USER_ZQY",
    "USER_ZGL",
    "allowed_skill_types_for_bot",
    "bot_key_for_bot_id",
    "can_user_access_bot",
    "is_known_bot_id",
    "resolve_bot_id",
    "resolve_permission_user_id",
]

File: apps/gateway/src/security/permission-policy.ts
import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join, resolve } from "node:path";
import { parseDocument } from "yaml";
import { repositoryRoot } from "@lxe/core";

export const ALL = "*";
export const POLICY_PATH_ENV = "LXE_PERMISSION_POLICY_PATH";
export class PermissionPolicyError extends Error {}

export interface PermissionPolicy {
  path: string;
  botIdToKey: ReadonlyMap<string, string>;
  botAliasToKey: ReadonlyMap<string, string>;
  botAliasToAppId: ReadonlyMap<string, string>;
  botSkillPolicy: ReadonlyMap<string, ReadonlySet<string>>;
  userAgentPolicy: ReadonlyMap<string, ReadonlySet<string>>;
  userNameToUnionId: ReadonlyMap<string, string>;
  userNameToAllowAliases: ReadonlyMap<string, ReadonlySet<string>>;
}

type Mapping = Record<string, unknown>;
const clean = (value: unknown): string => {
  if (value === null || value === undefined || value === false || value === 0 || value === "") return "";
  if (value === true) return "True";
  if (Array.isArray(value) && value.length === 0) return "";
  if (typeof value === "object" && Object.keys(value).length === 0) return "";
  return String(value).trim();
};
const isMapping = (value: unknown): value is Mapping =>
  value !== null && typeof value === "object" && !Array.isArray(value);

const requireMapping = (value: unknown, context: string): Mapping => {
  if (!isMapping(value)) throw new PermissionPolicyError(`${context} must be a mapping`);
  return value;
};

const stringList = (value: unknown, context: string): string[] => {
  if (!Array.isArray(value)) throw new PermissionPolicyError(`${context} must be a list`);
  const result: string[] = [];
  const seen = new Set<string>();
  for (const item of value) {
    const safe = clean(item);
    if (!safe) throw new PermissionPolicyError(`${context} contains an empty value`);
    if (seen.has(safe)) throw new PermissionPolicyError(`${context} contains duplicate value: ${safe}`);
    seen.add(safe);
    result.push(safe);
  }
  return result;
};

export function buildPermissionPolicy(data: unknown, path: string): PermissionPolicy {
  const root = requireMapping(data, "permission policy root");
  const bots = requireMapping(root.bots, "bots");
  const users = requireMapping(root.users, "users");
  if (Object.keys(bots).length === 0) throw new PermissionPolicyError("bots must not be empty");

  const botIdToKey = new Map<string, string>();
  const botAliasToKey = new Map<string, string>();
  const botAliasToAppId = new Map<string, string>();
  const botSkillPolicy = new Map<string, ReadonlySet<string>>();
  for (const [rawAlias, rawBot] of Object.entries(bots)) {
    const alias = clean(rawAlias);
    if (!alias) throw new PermissionPolicyError("bot alias must not be empty");
    const bot = requireMapping(rawBot, `bot ${alias}`);
    const key = clean(bot.key);
    const appId = clean(bot.app_id);
    const skills = stringList(bot.skill_types, `bot ${alias}.skill_types`);
    if (!key) throw new PermissionPolicyError(`bot ${alias}.key must not be empty`);
    if (!appId) throw new PermissionPolicyError(`bot ${alias}.app_id must not be empty`);
    if (skills.length === 0) throw new PermissionPolicyError(`bot ${alias}.skill_types must not be empty`);
    if (botIdToKey.has(appId)) throw new PermissionPolicyError(`duplicate bot app_id: ${appId}`);
    const skillSet = new Set(skills);
    const existing = botSkillPolicy.get(key);
    if (existing && (existing.size !== skillSet.size || [...existing].some((item) => !skillSet.has(item)))) {
      throw new PermissionPolicyError(`bot ${alias}.skill_types must match shared permission key: ${key}`);
    }
    botIdToKey.set(appId, key);
    botAliasToKey.set(alias, key);
    botAliasToAppId.set(alias, appId);
    botSkillPolicy.set(key, skillSet);
  }

  const userAgentPolicy = new Map<string, ReadonlySet<string>>();
  const userNameToUnionId = new Map<string, string>();
  const userNameToAllowAliases = new Map<string, ReadonlySet<string>>();
  for (const [rawName, rawUser] of Object.entries(users)) {
    const name = clean(rawName);
    if (!name) throw new PermissionPolicyError("user name must not be empty");
    const user = requireMapping(rawUser, `user ${name}`);
    const unionId = clean(user.union_id);
    const aliases = stringList(user.allow, `user ${name}.allow`);
    if (!unionId) throw new PermissionPolicyError(`user ${name}.union_id must not be empty`);
    if (aliases.length === 0) throw new PermissionPolicyError(`user ${name}.allow must not be empty`);
    if (userAgentPolicy.has(unionId)) throw new PermissionPolicyError(`duplicate user union_id: ${unionId}`);
    if (aliases.includes(ALL) && aliases.length > 1) {
      throw new PermissionPolicyError(`user ${name}.allow cannot mix '*' with bot aliases`);
    }
    const allowed = new Set<string>();
    for (const alias of aliases) {
      if (alias === ALL) allowed.add(ALL);
      else {
        const key = botAliasToKey.get(alias);
        if (!key) throw new PermissionPolicyError(`user ${name}.allow references unknown bot alias: ${alias}`);
        allowed.add(key);
      }
    }
    userAgentPolicy.set(unionId, allowed);
    userNameToUnionId.set(name, unionId);
    userNameToAllowAliases.set(name, new Set(aliases));
  }
  return {
    path,
    botIdToKey,
    botAliasToKey,
    botAliasToAppId,
    botSkillPolicy,
    userAgentPolicy,
    userNameToUnionId,
    userNameToAllowAliases,
  };
}

interface PermissionPolicyPathOptions {
  path?: string;
  env?: Readonly<Record<string, string | undefined>>;
  projectRoot?: string;
  home?: string;
}

const expandUser = (path: string, home: string): string => {
  if (path === "~") return home;
  if (path.startsWith("~/") || path.startsWith("~\\")) return join(home, path.slice(2));
  return path;
};

export function permissionPolicyPath(options: PermissionPolicyPathOptions = {}): string {
  const env = options.env ?? process.env;
  const configured = clean(options.path) || clean(env[POLICY_PATH_ENV]);
  if (configured) return expandUser(configured, options.home ?? homedir());
  const projectRoot = options.projectRoot ?? repositoryRoot(import.meta.dirname);
  return join(projectRoot, "config", "permission_policy.yaml");
}

export function loadPermissionPolicy(path?: string): PermissionPolicy {
  const safePath = permissionPolicyPath(path === undefined ? {} : { path });
  let content: string;
  try {
    content = readFileSync(safePath, "utf8");
  } catch (error) {
    const code = error instanceof Error && "code" in error ? String(error.code) : "";
    if (code === "ENOENT") throw new PermissionPolicyError(`permission policy file not found: ${safePath}`);
    throw new PermissionPolicyError(`cannot read permission policy file: ${safePath}: ${String(error)}`);
  }
  const document = parseDocument(content, { uniqueKeys: true });
  if (document.errors.length > 0) {
    const detail = document.errors[0]?.message ?? "unknown YAML error";
    const prefix = detail.includes("Map keys must be unique")
      ? "duplicate YAML key"
      : "invalid permission policy YAML";
    throw new PermissionPolicyError(`${prefix}: ${safePath}: ${detail}`);
  }
  return buildPermissionPolicy(document.toJS(), safePath);
}

export const botKeyForBotId = (policy: PermissionPolicy, botId: string): string =>
  policy.botIdToKey.get(clean(botId)) ?? "";
export const isKnownBotId = (policy: PermissionPolicy, botId: string): boolean =>
  Boolean(botKeyForBotId(policy, botId));
export function canUserAccessBot(policy: PermissionPolicy, userId: string, botId: string): boolean {
  const botKey = botKeyForBotId(policy, botId);
  if (!botKey) return false;
  const allowed = policy.userAgentPolicy.get(clean(userId)) ?? new Set<string>();
  return allowed.has(ALL) || allowed.has(botKey);
}

interface SourceLike {
  platform?: unknown;
  union_id?: unknown;
  source?: unknown;
  raw_data?: unknown;
}
const mapping = (value: unknown): Mapping => (isMapping(value) ? value : {});
export function resolvePermissionUserId(source: SourceLike): string {
  const raw = mapping(source.raw_data ?? source.source);
  return clean(source.union_id) || clean(raw.union_id) || clean(raw.sender_union_id);
}
export function resolveBotId(source: SourceLike, feishuAppId = ""): string {
  const raw = mapping(source.raw_data ?? source.source);
  const extra = mapping(raw.extra);
  const direct =
    clean(raw.bot_id) ||
    clean(raw.app_id) ||
    clean(raw.bot_app_id) ||
    clean(extra.bot_app_id) ||
    clean(extra.bot_id);
  const platform = (clean(source.platform) || clean(raw.platform)).toLowerCase();
  return platform === "feishu" ? direct || clean(feishuAppId) : direct;
}
